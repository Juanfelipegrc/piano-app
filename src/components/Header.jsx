import React from 'react'
import { usePianoStore } from '../hooks/usePianoStore'

export const Header = () => {

    const {onSetInstrument} = usePianoStore();

    const onChangeInstrument = ({target}) => {
        onSetInstrument(target.value)
    }

  return (
    <div className='container-fluid bg-black text-light p-3 d-flex justify-content-center align-items-center'>
        <h1 className='m-0 mr-2'>PIANO APP</h1>
        <form style={{height:'30px'}} className='flex-grow-1 ms-2'>
            <select className='form-select form-select-sm' name="instrument" id="instrument-select" onChange={onChangeInstrument}>
                <option value="acoustic_grand_piano">Piano</option>
                <option value="electric_guitar_jazz">Guitar</option>
                <option value="violin">Violin</option>
                <option value="trumpet">Trumpet</option>
                <option value="soprano_sax">Saxophone</option>
            </select>
        </form>

    </div>
  )
}
