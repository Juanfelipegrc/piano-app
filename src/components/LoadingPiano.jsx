import React from 'react'


export const LoadingPiano = ({onStart, isInitialized}) => {



    
  return (


    !isInitialized
    ?  <div
    className='container-fluid d-flex justify-content-center align-items-center container-loading-piano'

    >

        <button onClick={onStart} className='btn btn-success btn-lg'>
            <h1>Start playing</h1>
        </button>


    </div>

    : <> </>

   
  )
}
