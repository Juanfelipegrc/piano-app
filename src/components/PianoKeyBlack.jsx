import React from 'react'

export const PianoKeyBlack = ({note, playSound}) => {
  
    
    return (
        
        
        <div className='container-md'>
            {
                note != undefined?
                <div className='container'>
                <div 
                className='bg-black black-key'
                style={{
                    position: 'relative',
                    marginLeft: '8px',
                    width: '25px',
                    height: '100px',
                    color:'#fff',
                    zIndex:'100',
                    fontSize:'12px'

                    
                    }}
                    onMouseDown={(e) => {
                        e.stopPropagation();
                        playSound(note);
                      }}
                >
                    {note}
                </div>
            </div>

            :
                <></>

            }
     
    </div>
    
  )
}
