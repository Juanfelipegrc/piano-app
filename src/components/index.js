export * from './PianoKey';
export * from './PianoKeyBlack';
export * from './LoadingPiano';