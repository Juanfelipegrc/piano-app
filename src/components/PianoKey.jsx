import React, { useEffect, useState } from 'react';
import { PianoKeyBlack } from './PianoKeyBlack';

export const PianoKey = ({ notes, notesArray, playSound }) => {
  const [actualWidth, setActualWidth] = useState(window.innerWidth);
  

  const validate = (noteWhite, blackKeys) => {
    const noteBase = noteWhite.slice(0, -1); // Extract the note base (C, D, E, etc.)
    const noteOctave = noteWhite.slice(-1); // Extract the octave (3, 4, 5, etc.)
    return blackKeys.find(blackKey => 
      blackKey.includes(noteBase) && blackKey.includes(noteOctave)
    );
  };

  useEffect(() => {

    const handleResize = () => {
      setActualWidth(window.innerWidth);
    };

    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    }

  }, [])



  return (
    <div className={'h-100 d-flex justify-content-center align-items-center gap-0 flex-wrap'}>
      
      

        {
          notesArray.map((group, indexGroup) => {
            const whiteKeys = Object.keys(group).filter(note => !note.includes('#'));
            const blackKeys = Object.keys(group).filter(note => note.includes('#'));


       
            return (
              <div key={indexGroup} className='d-flex m-0 p-0 justify-content-center'>
                {
                  whiteKeys.map((noteWhite, index) => (
              
                    <div 
                      key={index}
                      className='bg-white position-relative'
                      style={{
                        width: '45px',
                        height: '200px',
                        border: 'solid 2px black',
                        margin: '0 1px'
                      
                      }}
                      onMouseDown={() => playSound(noteWhite)}
                    >
                      {noteWhite}
                      {
                        validate(noteWhite, blackKeys) && (
                          <PianoKeyBlack 
                            note={validate(noteWhite, blackKeys)}
                            playSound={playSound}
                            
                          />
                        )
                      }
                    </div>
                  ))
                  }
                
              </div>
            )
          })

        }
            
          
    </div>
  );
};
