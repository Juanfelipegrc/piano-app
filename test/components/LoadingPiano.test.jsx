import { fireEvent, render, screen } from "@testing-library/react"
import { LoadingPiano } from "../../src/components/LoadingPiano"


describe('Pruebas en <LoadingPiano>', () => { 

    test('debe de renderizar el botón y el isInitialized debe de estar en false', () => { 

        const {component} = render(<LoadingPiano onStart={() => {}} isInitialized={false}/>)

        expect(screen.getByText('Start playing')).toBeTruthy();

        expect(screen.getByRole('button')).toBeTruthy();

        expect(component).toMatchSnapshot();

     });

     test('no debe de renderizar el componente si el isInitialized está en true', () => { 

        render(<LoadingPiano onStart={() => {}} isInitialized={true}/>)

        expect(screen.queryByText('Start playing')).toBeNull();

        expect(screen.queryByRole('button')).toBeNull();

     });


     test('debe de llamar el onStart cuando se de click en el botón', () => { 
        const mockOnStart = jest.fn();
        render(<LoadingPiano onStart={mockOnStart} isInitialized={false}/>)

        const button = screen.getByRole('button');

        fireEvent.click(button);

        expect(mockOnStart).toHaveBeenCalled();

     });

     

 })