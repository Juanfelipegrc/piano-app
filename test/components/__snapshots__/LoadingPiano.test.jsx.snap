// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoadingPiano> debe de renderizar el bot√≥n y el isInitialized debe de estar en false 1`] = `undefined`;
